### PR TITLE
zio: Rework Attributes to be Descriptions

### PR DESCRIPTION
**Pull request by teburd**
Rather than providing the attribute values directly in get_attrs
for zio devices and channels, provide descriptions of each attribute.

This should allow a driver to map many attributes directly to registers
or internalized struct members and avoid storing these values in a
zio_variant which is confusing.

Signed-off-by: Tom Burdick <thomas.burdick@gmail.com>

### CODE REVIEW

**Comment by microbuilder** [comment]
@bfrog Just looking through this, but it might make sense to have two commits to review, one for the core zio API changes and the second for the synth driver updates. The driver is small enough it's not hard to review together, but maybe worth doing like that for other reviewers in the future. I do agree attributes seem a bit cumbersome as-is, though, at least on reasonably complex devices.

File: include/zio/zio_dev.h [hunk 1]
```diff
@@ -141,6 +147,37 @@ struct zio_dev_chan {
...
+ * @brief Pre-defined attribute types with the ability to extend by a driver
+ * implementation.
+ *
+ * Extended values should always be based on ZIO_ATTR_TYPES count with an
+ * additive and defined in a publicly available driver header.
+ *
+ * Example: #define MY_ATTR_TYPE (ZIO_ATTR_TYPES+1)
+ */
+enum zio_attr_type {
+	/* samples per second */
```
Comment (microbuilder) [inline include/zio/zio_dev.h]: Enum records generally have the comment **after** the field, with `/**<` to indicate that the comment block refers to the preceding field, so:

```
ZIO_SAMPLE_RATE,
/**< @brief samples per second */
...
```

File: include/zio/zio_dev.h [hunk 2]
```diff
@@ -141,6 +147,37 @@ struct zio_dev_chan {
...
+	/* enforce a specific enum size allowing for a good number of user
+	 * definable values but keeping
+	 */
+	ZIO_ATTR_TYPE_MAKE_16_BIT = 0xFFFF
+};
+
+/** Attribute record. */
+struct zio_attr_desc {
+	/* Attribute type, describing what it is such as the sample rate */
+	enum zio_attr_type type;
```
Comment (microbuilder) [inline include/zio/zio_dev.h]: Doxygen comment blocks need to start with `**`, and would benefit from the `@brief` tag before the text.

Struct attributes generally have the comment **after** the field, though, with `/**<` to indicate that the comment block refers to the preceding field, so:

```
enum zio_attr_type type;
/**< @brief Attribute type, describing what it is such as the sample rate */

...
```

**Review by microbuilder — CHANGES_REQUESTED** [review]
Looks good other than some doxygen issues commented inline.

File: include/zio/zio_dev.h [hunk 3]
```diff
@@ -80,14 +79,21 @@ struct zio_buf;
  * @brief Pre-defined channel types with the ability to extend by a driver
  * implementation.
  *
- * Extended values should always be based on ZIO_LAST_CHAN_TYPE with an
+ * Extended values should always be based on ZIO_CHAN_TYPES count with an
  * additive and defined in a publicly available driver header.
+ *
+ * Example: #define MY_CHAN_TYPE (ZIO_CHAN_TYPES+1)
```
Comment (dbkinder) [inline include/zio/zio_dev.h]: ``#name`` in a doxygen comment is used to reference a variable named "name", hence the error that's showing up saying:
```
warning: explicit link request to 'define' could not be resolved
```
Workaround is to escape the ``#`` with ``\#``

```suggestion
 * Example: \#define MY_CHAN_TYPE (ZIO_CHAN_TYPES+1)
```

File: include/zio/zio_dev.h [hunk 4]
```diff
@@ -129,18 +135,52 @@ struct zio_dev_chan {
...
 };
 
+/**
+ * @brief Pre-defined attribute types with the ability to extend by a driver
+ * implementation.
+ *
+ * Extended values should always be based on ZIO_ATTR_TYPES count with an
+ * additive and defined in a publicly available driver header.
+ *
+ * Example: #define MY_ATTR_TYPE (ZIO_ATTR_TYPES+1)
```
Comment (dbkinder) [inline include/zio/zio_dev.h]:
```suggestion
 * Example: \#define MY_ATTR_TYPE (ZIO_ATTR_TYPES+1)
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
fix doc warnings from doxygen comments

**Comment by bfrog** [comment]
I believe I've address all the concerns @dbkinder @microbuilder

File: doc/reference/zio/index.rst [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ZIO Device API
+**************
+
+.. doxygengroup:: zio_device
+   :project: Zephyr
+
+ZIO Attribute Definitions
+*************************
+
+.. doxygengroup:: zio_attributes
```
Comment (dbkinder) [inline doc/reference/zio/index.rst]: This doxygen group isn't defined (in this PR)

**Review by dbkinder — CHANGES_REQUESTED** [review]
When I pull your PR locally, I see a file ``samples/zio/synth/README.rst that doesn't appear to actually part of your PR file list.  Not sure what's up (maybe a rebase is needed)?

File: doc/reference/zio/index.rst (hunk 5 above)
Comment (bfrog) [reply doc/reference/zio/index.rst]: doh, thanks, will fix this up

Comment (bfrog) [reply doc/reference/zio/index.rst]: I've removed that doxygen group for now

**Comment by bfrog** [comment]
@dbkinder I'm still not sure what the documentation problems here are exactly, any clues would be helpful

**Review by microbuilder — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]
@dbkinder please revisit

**Review by dbkinder — APPROVED** [review]
I'm not seeing any doc build errors when I generate docs locally, so LGTM.

**Comment by MaureenHelm** [comment]
> I'm not seeing any doc build errors when I generate docs locally, so LGTM.

Thanks!

**Pull request merged**